Fold SVE mul and mul_u to neg during isel

MartinWehking opened the pull request:
Replace mul and mul_u ops with a neg operation if their second operand is a splat value -1.

Apply the optimization also for mul_u ops if their first operand is a splat value -1 due to their commutativity.

MartinWehking committed:
Fold SVE mul and mul_u to neg during isel

Replace mul and mul_u ops with a neg operation if their second operand
is a splat value -1.

Apply the optimization also for mul_u ops if their first operand
is a splat value -1 due to their commutativity.

Lukacma commented on llvm/lib/Target/AArch64/SVEInstrFormats.td:
```diff
@@ -723,6 +723,19 @@ class SVE2p1_Cvt_VG2_Pat<string name, SDPatternOperator intrinsic, ValueType out
 
+class SVE_2_Op_Neg_One_Passthru_Pat<ValueType vt, SDPatternOperator op, ValueType pt,
+                             Instruction inst, ValueType immT>
+: Pat<(vt (op pt:$Op1, vt:$Op2, (vt (splat_vector (immT -1))))),
```
I think that similarly to AArch64bic we need to extend this to work for both 32 and 64 bit splats. Or is there a reason why you are enabling it only for one or the other ?

MartinWehking replied:
Hi, thanks, that's something which has made me think about it a lot.

I don't think it's needed here. AArch64bic is a PatFrags that is used generically for all patterns that might occur. As far as I know, only for a splat vector of type "nxv2i64", the 64 bit splat is required.

I created multiple patterns where the splat value dataype is provided as a parameter.
Similar to other patterns (e.g. the ones listed under "// SVE2 integer add/subtract wide "  (and the SelectionDAG I saw) the type is either 32bit for the immediate when the result vector is not  "nxv2i64" or 64bit when it is.
I'm assuming due to type legalization there won't be any e.g. i16 bit type.

If there is the possibility for initializing a splat_vector is wrong and there is a valid 64 bit splat for e.g. a non `nxv2i64` then this needs to be adapted. Please also correct me if there is a misunderstanding from my side.

Lukacma commented on llvm/test/CodeGen/AArch64/sve-int-mul-neg.ll:
```diff
@@ -0,0 +1,213 @@
+  %2 = call <vscale x 2 x i64> @llvm.aarch64.sve.mul.u.nxv2i64(<vscale x 2 x i1> %pg, <vscale x 2 x i64> %1, <vscale x 2 x i64> %a)
+  ret <vscale x 2 x i64> %2
+}
+; Non foldable muls -- we don't expect these to be optimised out.
```
I dont think these negative tests are necessary here.

Lukacma commented on llvm/test/CodeGen/AArch64/sve-int-mul-neg.ll:
```diff
@@ -0,0 +1,213 @@
+; CHECK-NEXT:    mov z0.d, z1.d
+; CHECK-NEXT:    ret
+  %1 = call <vscale x 2 x i64> @llvm.aarch64.sve.dup.x.nxv2i64(i64 -1)
+  %2 = call <vscale x 2 x i64> @llvm.aarch64.sve.mul.nxv2i64(<vscale x 2 x i1> %pg, <vscale x 2 x i64> %1, <vscale x 2 x i64> %a)
```
I think this is missed optimization opportunity. We could generate:
```
mov z1.d, -1
neg z1.d, p0/m, z0
```

which is better. So we should probably add patterns for these cases as well.

MartinWehking committed:
Lower to undef variants of neg for _u variants

Target undef variants for variants of mul where the inactive lanes do
not matter (mul.u instrinsics).

Create also test cases for all different datatypes when checking
the commutativity of mul.u.

Remove target guards since only optimization patterns are
added.

MartinWehking commented on llvm/test/CodeGen/AArch64/sve-int-mul-neg.ll:
```diff
@@ -0,0 +1,213 @@
+; CHECK-NEXT:    neg z0.d, p0/m, z0.d
+; CHECK-NEXT:    ret
+  %1 = call <vscale x 2 x i64> @llvm.aarch64.sve.dup.x.nxv2i64(i64 -1)
+  %2 = call <vscale x 2 x i64> @llvm.aarch64.sve.mul.u.nxv2i64(<vscale x 2 x i1> %pg, <vscale x 2 x i64> %1, <vscale x 2 x i64> %a)
```
Thanks, I added them

MartinWehking committed:
Add different argument order pattern for mul

Optimize mul for merge variants where the splat value is the first
operand.
This optimization is equal to a neg that as a (mov -1) and op2 as
operands.

Remove a redundant commutativity pattern for undef variants of mul since
the splat value should appear only as second operand after instcombine.

Adapt the tests to test for the new patterns and remove redundant ones
for non-optimization cases.

MartinWehking commented on llvm/lib/Target/AArch64/SVEInstrFormats.td:
```diff
@@ -723,6 +723,16 @@ class SVE2p1_Cvt_VG2_Pat<string name, SDPatternOperator intrinsic, ValueType out
     : Pat<(out_vt (intrinsic in_vt:$Zn1, in_vt:$Zn2)),
                   (!cast<Instruction>(name) (REG_SEQUENCE ZPR2Mul2, in_vt:$Zn1, zsub0, in_vt:$Zn2, zsub1))>;
 
+class SVE_2_Op_Neg_One_Replace_Pat<ValueType vt, SDPatternOperator op,  ValueType pt,
```
Could we get another opinion by one of the other reviewers on this?
@CarolineConcatto @paulwalker-arm  

Thanks

paulwalker-arm replied:
Now that each class only contains a single pattern and the patterns themselves are specific to a single instruction I agree with @Lukacma that it will be cleaner to have the patterns inlined into `AArch64SVEInstrInfo.td`.

MartinWehking committed:
Define Patterns directly

Do not use helper classes and create the patterns directly

Move pattern definitions into a similar spot

Lukacma reviewed (approved):
LGTM

paulwalker-arm commented on llvm/test/CodeGen/AArch64/sve-int-mul-neg.ll:
```diff
@@ -0,0 +1,148 @@
+; CHECK:       // %bb.0:
+; CHECK-NEXT:    neg z0.b, p0/m, z0.b
+; CHECK-NEXT:    ret
+  %1 = call <vscale x 16 x i8> @llvm.aarch64.sve.dup.x.nxv16i8(i8 -1)
```
Not that important but you could simplify the tests by removing all the dup.x calls and use `splat(<typ> -1)` inline at use.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64SVEInstrInfo.td:
```diff
@@ -1012,6 +1012,36 @@ let Predicates = [HasSVE_or_SME] in {
+  }
+
+  def : Pat<(nxv16i8 (AArch64mul_m1 nxv16i1:$Op1, (nxv16i8 (splat_vector (i32 -1))), nxv16i8:$Op2)),
+        (NEG_ZPmZ_B (DUP_ZI_B -1, 0) , $Op1, $Op2)>;
```
Rouge whitespace `(DUP_ZI_B -1, 0) ,`.

paulwalker-arm reviewed (approved):
A few suggestions but otherwise this looks good to me.

MartinWehking committed:
Simplify test cases and reformat patch

Replace some dups with splat_vectors and remove whitespaces.
Remove redundant arg in run line.
